### PR TITLE
add seccomp.IsEnabled() function

### PR DESCRIPTION
**Pull request by jfrazelle**
This is much like apparmor.IsEnabled() function and a nice helper.

Signed-off-by: Jessica Frazelle <acidburn@docker.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
There is a risk of the process getting killed when trying to call make this system call. We could use the procfs to get the same information. From the man page
```
prctl(2)
Since Linux 3.8, the Seccomp field of the /proc/[pid]/status
              file provides a method of obtaining the same information,
              without the risk that the process is killed; see proc(5).

proc(5)
/proc/[pid]/seccomp (from Linux 2.6.12 to 2.6.22)
              Read/set the seccomp mode for the process.  If this file
              contains the value zero, seccomp mode is not enabled.  Writing
              the value 1 to this file (irreversibly) places the process in
              seccomp mode: the only permitted system calls are read(2),
              write(2), _exit(2), and sigreturn(2).  This file went away in
              Linux 2.6.23, when it was replaced by a prctl(2)-based
              mechanism.

```

**Comment by crosbymichael** [comment]
@mrunalp how could it get killed?

**Comment by jfrazelle** [comment]
it would only kill it if you have seccomp and have denied prctl

**Comment by cyphar** [comment]
Since we require a fairly modern version of the kernel, we should at least attempt to read from `/proc/self/status` before we use `prctl`. That way, at least modern clients won't be killed when there's a better way.

WDYT @jfrazelle?470

**Comment by jfrazelle** [comment]
its not going to work if they don't have it compiled in their kernel it
won't show up in /proc/pid... also we can't turn on seccomp on a pid if its
not compiled in the kernel and we can't check it to make sure it's compiled
in the kernel without prctl, its kinda a problem w order of operations

**Comment by cyphar** [comment]
@jfrazelle Wouldn't something like this be okay for us:

1. `state=$(grep '^Seccomp:' /proc/self/status | awk '{ print $2 }'`
2. If no match found, do `prctl` to double-check (either it's not compiled into the kernel or they're on `linux<3.8`).

**Comment by mrunalp** [comment]
I agree with @cyphar to use prctl as a fallback.

**Comment by jfrazelle** [comment]
will update !

**Comment by crosbymichael** [comment]
No one ever said why or how the prctl will fail?

**Comment by jfrazelle** [comment]
it will kill the process if the process is running with seccomp and the
seccomp filters defined prctl as kill errno, etc

**Comment by crosbymichael** [comment]
But this is not going to be running inside a container.  its run on the a daemon/upper layers to see if seccomp is supported.

**Comment by jfrazelle** [comment]
Ya so we should never hit it

**Comment by mrunalp** [comment]
@crosbymichael Agree, if we never plan to use this in nested containers. However, to make it more useful, it makes sense to make it work for most scenarios. However, if you feel strongly that we shouldn't do that now, I am okay with it and adding reading from /proc/pid/status later on as required.

**Comment by jfrazelle** [comment]
updated ptal

File: libcontainer/seccomp/seccomp_linux.go [hunk 1]
```diff
@@ -178,3 +201,28 @@ func matchCall(filter *libseccomp.ScmpFilter, call *configs.Syscall) error {
...
+	s := bufio.NewScanner(f)
+	status := make(map[string]string)
+
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return nil, err
+		}
+
+		text := s.Text()
+		parts := strings.Split(text, ":")
```
Comment (crosbymichael) [inline libcontainer/seccomp/seccomp_linux.go]: maybe check `len()` so that we don't panic?

**Comment by crosbymichael** [comment]
@mrunalp ok, thanks for the explanation

**Comment by jfrazelle** [comment]
adding a unit text too, 1 sec

File: libcontainer/seccomp/seccomp_linux.go [hunk 2]
```diff
@@ -73,6 +79,23 @@ func InitSeccomp(config *configs.Seccomp) error {
 	return nil
 }
 
+// IsEnabled returns if the kernel has been configured to support seccomp.
+func IsEnabled() bool {
+	// Try to read from /proc/self/status for kernels > 3.8
+	s, _ := parseStatusFile("/proc/self/status")
```
Comment (crosbymichael) [inline libcontainer/seccomp/seccomp_linux.go]: Need an error check here because you `return nil, err` in the function.

**Comment by mrunalp** [comment]
The second code path using prctl also checks for `CONFIG_SECCOMP_FILTER` which the first doesn't. Not sure whether we want to resolve that (or if it's even possible). Otherwise LGTM.

**Comment by crosbymichael** [comment]
LGTM

@mrunalp we fixed the fall-through issue.

**Pull request merged**